Custom TypeScript types for better RDF/JS compatibility

minorg opened the pull request:
My goal was to make the the types generated by wasm more compatible with [RDF/JS Dataset types](https://rdf.js.org/dataset-spec/). For example, term implementations are expected to have a string literal `termType` and not a `string`. The TypeScript compiler complains about this.

Rather than writing ugly wrappers in my TypeScript, I tried to fix the generated `.d.ts` here. In `model.rs` and `store.rs` I marked the term classes, the Store class, and two functions to `skip_typescript` in `wasm_bindgen`, then provided a `typescript_custom_section` with more specific types, per the approach suggested [in this comment](https://github.com/rustwasm/wasm-bindgen/issues/1591#issuecomment-1705735439).

For reference, the RDF/JS reference types are in [data-model.d.ts](https://github.com/rdfjs/types/blob/master/data-model.d.ts) and [dataset.d.ts](https://github.com/rdfjs/types/blob/master/dataset.d.ts).

I did my best with the `Store` class by analyzing the code. I'm relatively inexperienced with Rust. Corrections would obviously be welcome. I didn't try to make the Rust `Store` methods return `this`, since `Store` isn't a complete `DatasetCore` anyway without a quad iterator implementation.

The resulting `.d.ts`:

```
/* tslint:disable */
/* eslint-disable */
/**
* @param {string} value
* @returns {NamedNode}
*/
export function namedNode(value: string): NamedNode;
/**
* @param {string | undefined} [value]
* @returns {BlankNode}
*/
export function blankNode(value?: string): BlankNode;
/**
* @returns {DefaultGraph}
*/
export function defaultGraph(): DefaultGraph;
/**
* @param {string} value
* @returns {Variable}
*/
export function variable(value: string): Variable;
/**
* @param {any} subject
* @param {any} predicate
* @param {any} object
* @returns {Quad}
*/
export function triple(subject: any, predicate: any, object: any): Quad;
/**
* @param {any} original
* @returns {any}
*/
export function fromTerm(original: any): any;
/**
* @param {any} original
* @returns {any}
*/
export function fromQuad(original: any): any;
/**
*/
export function main(): void;

export function literal(value: string | undefined, languageOrDataType?: string | NamedNode): Literal;
export function quad(subject: Quad_Subject, predicate: Quad_Predicate, object: Quad_Object, graph: Quad_Graph): Quad;

export class BlankNode {
    readonly termType: "BlankNode";
    readonly value: string;

    equals(other: Term | null | undefined): boolean;
    toString(): string;
}

export class DefaultGraph {
    readonly termType: "DefaultGraph";
    readonly value: "";

    equals(other: Term | null | undefined): boolean;
    toString(): string;
}

export class Literal {
    readonly datatype: NamedNode;
    readonly language: string;
    readonly termType: "Literal";
    readonly value: string;

    equals(other: Term | null | undefined): boolean;
    toString(): string;
}

export class NamedNode {
    readonly termType: "NamedNode";
    readonly value: string;

    equals(other: Term | null | undefined): boolean;
    toString(): string;
}

export interface BaseQuad {
    readonly graph: Term;
    readonly object: Term;
    readonly predicate: Term;
    readonly subject: Term;
    readonly termType: "Quad";
    readonly value: "";

    equals(other: Term | null | undefined): boolean;
    toString(): string;
}

export class Quad implements BaseQuad {
    readonly graph: Quad_Graph;
    readonly object: Quad_Object;
    readonly predicate: Quad_Predicate;
    readonly subject: Quad_Subject;
    readonly termType: "Quad";
    readonly value: "";

    equals(other: Term | null | undefined): boolean;
    toString(): string;
}

export type Quad_Graph = DefaultGraph | NamedNode | BlankNode | Variable;
export type Quad_Object = NamedNode | Literal | BlankNode | Quad | Variable;
export type Quad_Predicate = NamedNode | Variable;
export type Quad_Subject = NamedNode | BlankNode | Quad | Variable;

export type Term = NamedNode | BlankNode | Literal | Variable | DefaultGraph | BaseQuad;

export class Variable {
    readonly termType: "Variable";
    readonly value: string;

    equals(other: Term | null | undefined): boolean;
    toString(): string;
}



export class Store {
    readonly size: number;

    constructor(quads?: Quad[]);

    add(quad: Quad): void;

    delete(quad: Quad): void;

    dump(
        options: {
            format: string;
            from_graph_name?: BlankNode | DefaultGraph | NamedNode;
        },
        from_graph_name?: BlankNode | DefaultGraph | NamedNode
    ): string;

    has(quad: Quad): boolean;

    load(
        data: string,
        options: {
            base_iri?: NamedNode | string;
            format: string;
            no_transaction?: boolean;
            to_graph_name?: BlankNode | DefaultGraph | NamedNode;
        },
        base_iri?: NamedNode | string,
        to_graph_name?: BlankNode | DefaultGraph | NamedNode
    ): void;

    match(subject?: Term | null, predicate?: Term | null, object?: Term | null, graph?: Term | null): Quad[];

    query(
        query: string,
        options?: {
            base_iri?: NamedNode | string;
            results_format?: string;
            use_default_graph_as_union?: boolean;
        }
    ): boolean | Map<string, Term>[] | Quad[];

    update(
        update: string,
        options?: {
            base_iri?: NamedNode | string;
        }
    ): void;
}
```

Tpt reviewed (commented):
Thank you so much! This is very welcome! It might be great to convert the tests to typescript in order to have some way to check the type stubs. But this is not for this MR.

I allowed myself some nitpicks, will merge the PR as soon as they are fixed (or if you prefer I can merge now and do them in a follow up PR)

Tip: `cargo fmt` should make the "fmt" lint happy

minorg commented on js/src/model.rs:
```diff
@@ -10,6 +10,85 @@ thread_local! {
+
+    equals(other: Term | null | undefined): boolean;
+    toString(): string;
+}
```
I left it out initially, but:
* `DatasetCore` `match` method takes `Term` parameters
* The RDF/JS `Term` unions `BaseQuad` instead of `Quad`

So there's type incompatibility if `match` takes `Quad` instead of `BaseQuad`.
I admit I don't really understand the rationale behind that, and it doesn't seem to be documented in sources I could find.

I copied `BaseQuad`, the `Term` union, and the `match` signature from the RDF/JS types.

Tpt replied:
Ok! Make sense. Thank you! Maybe add a comment?

minorg replied:
I added comments to the generated TypeScript and a headline comment to the Rust.

minorg commented on js/src/store.rs:
```diff
@@ -9,7 +9,61 @@ use oxigraph::sparql::{Query, QueryResults, Update};
+            format: string;
+            from_graph_name?: BlankNode | DefaultGraph | NamedNode;
+        },
+        from_graph_name?: BlankNode | DefaultGraph | NamedNode
```
I saw that in the Rust but wasn't sure about allowing it here. I'll change it.

minorg commented:
@Tpt I wondered why `wasm-pack` is not in the `package.json` `devDependencies`? I added it in my local checkout so the build script would work but didn't commit my changes.

minorg commented:
@Tpt I went ahead and changed the tests to TypeScript, which revealed a couple of type issues in the custom TypeScript declarations. I was concerned about that with the initial PR. I'm glad you nudged me in that direction.

I changed the tests to use vitest instead of Mocha. It handles TypeScript a little more gracefully. I enabled defineGlobals in the vitest config so the rdfjs-datamodel Mocha tests (which assume global describe and it) still run.

Tpt commented on js/tsconfig.json:
```diff
@@ -0,0 +1,13 @@
+        "experimentalDecorators": true,
+        "forceConsistentCasingInFileNames": true,
+        "noUncheckedIndexedAccess": false,
+        "incremental": true
```
Genuine question (I am not a TypeScript expert): do we need all of these if we already extend node18 and strictest config?

minorg replied:
Probably not. I carried them from another project.
I've removed them all and fixed a bit of test code that assumed "noUncheckedIndexAccess": false.

Tpt reviewed (commented):
Thank you for the improvements. The migration to vitest is very welcome, it will make benchmarking easier ([codespeed has a great vitest integration](https://docs.codspeed.io/benchmarks/nodejs/vitest))

Sadly CI seems to be broken because [of a limitation of getrandom with ES modules](https://docs.rs/getrandom/latest/getrandom/#nodejs-es-module-support)).

minorg commented on js/src/store.rs:
```diff
@@ -9,7 +9,61 @@ use oxigraph::sparql::{Query, QueryResults, Update};
+            results_format?: string;
+            use_default_graph_as_union?: boolean;
+        }
+    ): boolean | Map<string, Term>[] | Quad[];
```
I changed this.

minorg commented:
> Sadly CI seems to be broken because [of a limitation of getrandom with ES modules](https://docs.rs/getrandom/latest/getrandom/#nodejs-es-module-support)).

I found the right incantation to work around this:

```
import { webcrypto } from "node:crypto";
vi.stubGlobal("crypto", webcrypto);
```
where `vi` is imported from `vitest`.

Tpt reviewed (approved):
Thank you!
